kexoqs.c

```diff
@@ -480,51 +324,50 @@ int
 oqs_server_gen_msg_and_ss(OQS_KEX_CTX *oqs_kex_ctx,
 	u_char **oqs_shared_secret, size_t *oqs_shared_secret_len) {
 
-	OQS_KEX *oqs_kex = NULL;
+	OQS_KEM *oqs_kem = NULL;
 	uint8_t *tmp_oqs_shared_secret = NULL;
-	size_t tmp_oqs_shared_secret_len = 0;
 	int r = 0;
 
-	if (OQS_NEED_SEED == oqs_kex_ctx->oqs_need_seed) {
-		if (oqs_kex_ctx->oqs_seed == NULL ||
-			oqs_kex_ctx->oqs_seed_len != oqs_kex_ctx->oqs_need_seed_len) {
-			fprintf(stderr, "seed failed\n");
-			r = SSH_ERR_INTERNAL_ERROR;
-			goto out;
-		}
+	if ((oqs_kem = OQS_KEM_new(oqs_kex_ctx->oqs_method)) == NULL) {
+		r = SSH_ERR_INTERNAL_ERROR;
+		goto out;
 	}
-	else {
-		oqs_kex_ctx->oqs_seed = NULL;
-		oqs_kex_ctx->oqs_seed_len = 0;
+
+	if ((oqs_kex_ctx->oqs_local_msg = malloc(oqs_kem->length_ciphertext)) == NULL) {
+		r = SSH_ERR_ALLOC_FAIL;
+		goto out;
+	}
+	if ((tmp_oqs_shared_secret = malloc(oqs_kem->length_shared_secret)) == NULL) {
+		r = SSH_ERR_ALLOC_FAIL;
+		goto out;
 	}
 
-	if ((oqs_kex = OQS_KEX_new(&oqs_rnd, oqs_kex_ctx->oqs_method,
-		oqs_kex_ctx->oqs_seed, oqs_kex_ctx->oqs_seed_len,
-		oqs_kex_ctx->oqs_param)) == NULL) {
-		r = SSH_ERR_INTERNAL_ERROR;
+	if (oqs_kex_ctx->oqs_remote_msg_len != oqs_kem->length_public_key) {
+		r = SSH_ERR_INVALID_FORMAT;
 		goto out;
 	}
 
-	if (OQS_KEX_bob(oqs_kex, oqs_kex_ctx->oqs_remote_msg,
-		oqs_kex_ctx->oqs_remote_msg_len, &(oqs_kex_ctx->oqs_local_msg),
-		&(oqs_kex_ctx->oqs_local_msg_len), &tmp_oqs_shared_secret,
-		&tmp_oqs_shared_secret_len) != OQS_SUCCESS) {
-		r = SSH_ERR_INTERNAL_ERROR;
+	if (OQS_KEM_encaps(oqs_kem, oqs_kex_ctx->oqs_local_msg, tmp_oqs_shared_secret,
+		oqs_kex_ctx->oqs_remote_msg) != OQS_SUCCESS) {
+				r = SSH_ERR_INTERNAL_ERROR;
 		goto out;
 	}
 
 	*oqs_shared_secret = (u_char *) tmp_oqs_shared_secret;
-	*oqs_shared_secret_len = tmp_oqs_shared_secret_len;
+	*oqs_shared_secret_len = oqs_kem->length_shared_secret;
+	oqs_kex_ctx->oqs_local_msg_len = oqs_kem->length_ciphertext;
 
 	tmp_oqs_shared_secret = NULL;
 
 out:
-	if (oqs_kex != NULL) {
-		OQS_KEX_free(oqs_kex);
+	if (oqs_kem != NULL) {
+		OQS_KEM_free(oqs_kem);
 	}
 	if (tmp_oqs_shared_secret != NULL) {
-		explicit_bzero(tmp_oqs_shared_secret, tmp_oqs_shared_secret_len);
+		explicit_bzero(tmp_oqs_shared_secret, oqs_kem->length_shared_secret);
 		free(tmp_oqs_shared_secret);
+		free(oqs_kex_ctx->oqs_local_msg);
```

Comment:
I would prefer having a tmp local message variable and then point the struct pointer to this if there are no errors.